### PR TITLE
Give ability to httpClient serialize/deserialize xml

### PR DESCRIPTION
**Pull request by svishnyakoff**
Add XML support for the http client. If server response or client request will have appropriate XML media type headers, then XmlMediaTypeCodec is going to be used for encoding/decoding. 

The PR is still not in its final state and some issue with for example dependencies have to be resolved but if overall Microunaut team would agree these changes might be a good candidate top be merged, I will work on all the cleanup work

### CODE REVIEW

**Comment by svishnyakov** [comment]
hey @graemerocher I have already signed it several types but status was not updated by some reason 
image

**Comment by jameskleeh** [comment]
@svishnyakov You made a commit with another account: "sergey.vishnyakov". That account has to accept the CLA

**Comment by svishnyakov** [comment]
thanks, that helped! At this point, the changes are ready for review. My main concern I would rather fixing is a transitive dependency on jackson.dataformat, that is better to be optional

**Comment by svishnyakov** [comment]
Seems my last changes broke tests so hold off the review until the build is green

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java [hunk 1]
```diff
@@ -677,13 +695,18 @@ private void handleRouteMatch(
...
-                    beanLocator.findBean(HttpContentSubscriberFactory.class,
-                        new ConsumesMediaTypeQualifier<>(type))
-                ).map(factory ->
-                    factory.build(request)
-                ).orElse(new DefaultHttpContentProcessor(request, serverConfiguration));
+            Optional<Class<?>> bodyType = route.getBodyArgument().map(Argument::getType);
+            boolean isFunctionalBean = route.getAnnotation("io.micronaut.function.FunctionBean") != null;
+
+            HttpContentProcessor<?> processor;
+            if (isFunctionalBean && bodyType.isPresent() && String.class.isAssignableFrom(bodyType.get())) {
```
Comment (jameskleeh) [inline http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java]: Why was this change necessary? Seems like a hack to work around another problem

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/xml/XmlContentProcessorSpec.groovy [hunk 2]
```diff
@@ -70,7 +84,6 @@ class XmlContentProcessorSpec extends Specification {
         then:
         author.books.size() == 1
         author.books[0].title == "First Book"
-        author.name == "Joe"
```
Comment (jameskleeh) [inline http-server-netty/src/test/groovy/io/micronaut/http/server/netty/xml/XmlContentProcessorSpec.groovy]: Why was this line removed?

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java (hunk 1 above)
Comment (svishnyakov) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java]: The current behaviour of how request deserialization works for FunctionBean is `really` inconsistent.
1. The documentation says that parameter should contain the entire body . 
> In addition, functions have an input and/or an output. The input is represented by the accepted argument and represents the body consumed by the function and the output is represented by the return value of the function. The input and the output should be either a Java primitive or simple type (int, String etc.) or a POJO.

2. For groovy based FunctionBean and with json request we bind json parameter to bean parameter

3. For java based FunctionBean binding of json fields does not work at all 

4. For xml request we have a test that also binding entire body to the parameter https://github.com/micronaut-projects/micronaut-core/blob/afc3fb377fea37f942c1a251be19b58f9f989b22/function-web/src/test/groovy/io/micronaut/function/web/JavaLambdaFunctionSpec.groovy#L113

I struggle to fix the tests, but due to these inconsistencies once I fix one of them, another gets broken.

File: http-server-netty/src/test/groovy/io/micronaut/http/server/netty/xml/XmlContentProcessorSpec.groovy (hunk 2 above)
Comment (svishnyakov) [reply http-server-netty/src/test/groovy/io/micronaut/http/server/netty/xml/XmlContentProcessorSpec.groovy]: Oops, will get it back

File: http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java (hunk 1 above)
Comment (jameskleeh) [reply http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java]: Yeah so I wouldn't hide any inconsistencies we have with a workaround. Let the tests be failing so we can have a clear understanding of what isn't working and then we can discuss how it should behave. Thoughts on the above @graemerocher ? I really haven't worked with the functions enough to determine how they should work.

**Comment by svishnyakov** [comment]
@jameskleeh @graemerocher I got to the stage when I am able to deserialize xml body as a java object and it should work well with `@Body` annotation when we just bind entire request to some param, BUT I am really stuck when we want to bind just some XML subfields to controller param

Examples that dont work yet 
```
<author name="Joe"><books><book><title>First Book</title></book><book><title>Second Book</title></book></books></author>'

        @Post("/author/books")
        Flowable<Book> bookList(List<Book> books) {
            return Flowable.fromIterable(books)
        }

        @Post("/author/bookSet")
        Flowable<Book> bookList(Set<Book> books) {
            return Flowable.fromIterable(books)
        }
```

Any advice what might be the next steps to fix param binding?

**Comment by svishnyakov** [comment]
@jameskleeh I fixed tests related to xml deseriziation when parameter marked with `@Body` and field injection also working now. Tests that actually are failing related to `@FunctionalBean` https://github.com/micronaut-projects/micronaut-core/pull/2045/files/b347a19a8a52d3db2341a9c5c5532da4278e9c4b#diff-0f91f8567f30707e6e437d24f3109b89, but besides them, everything else should be functional. Please help me with review

File: benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+public class XmlDeserializationBenchmark {
```
Comment (svishnyakov) [inline benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java]: @jameskleeh So here are some benchmark results. I have been running same tests against changes from current PR that  is using XmlMapper as a deserializer and changes from `xml-processing` branch that is converting xml stream to map-like structure that is later converted to object by Jackson. 

The benchmark results are not 100% fair cause `xml-processing` branch still lacks the support of collection deserialization that will require additional logic and usage of introspection functionality. So for the benchmark, I pick the xml structure that works in both branches just to get a general sense of performance degradation in case we use XmlMapper

#### That is interesting:
- Deserialization with introspected model work same or even worse results
- With map-like deserialization, on average it takes  0.015 ms/op; while with xml mapper it is 0.020. With xml-mapper it works a bit slower, but we get full XML support and JAXB would work too 

### Results from current PR (using XmlMapper)
https://gist.github.com/svishnyakov/46ea21d2d4824fcc3951dd17e0224073
Benchmark                                                                                                              Mode     Cnt   Score    Error   Units
XmlDeserializationBenchmark.streamingXmlDeserialization                                                               thrpt       4   6.531 ± 17.469  ops/ms
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                   thrpt       4   7.434 ± 14.289  ops/ms
XmlDeserializationBenchmark.xmlDeserialization                                                                        thrpt       4  48.473 ±  0.968  ops/ms
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                            thrpt       4  47.282 ±  1.216  ops/ms
XmlDeserializationBenchmark.streamingXmlDeserialization                                                              sample  408257   0.098 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.00                            sample           0.084            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.50                            sample           0.093            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.90                            sample           0.107            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.95                            sample           0.119            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.99                            sample           0.165            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.999                           sample           0.251            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.9999                          sample           1.531            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p1.00                            sample           4.612            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                  sample  394288   0.101 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.00    sample           0.085            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.50    sample           0.098            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.90    sample           0.109            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.95    sample           0.123            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.99    sample           0.177            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.999   sample           0.269            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.9999  sample           1.525            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p1.00    sample           3.600            ms/op
XmlDeserializationBenchmark.xmlDeserialization                                                                       sample  983241   0.020 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.00                                              sample           0.017            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.50                                              sample           0.019            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.90                                              sample           0.022            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.95                                              sample           0.022            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.99                                              sample           0.036            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.999                                             sample           0.073            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.9999                                            sample           1.278            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p1.00                                              sample           2.744            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                           sample  963280   0.021 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.00                      sample           0.018            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.50                      sample           0.020            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.90                      sample           0.022            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.95                      sample           0.023            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.99                      sample           0.035            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.999                     sample           0.069            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.9999                    sample           1.253            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p1.00                      sample           2.843            ms/op

### Results when converting to map-like structure:
https://gist.github.com/svishnyakov/00920090ac11dfa03ea93ebd6f0e320f
Benchmark                                                                                                              Mode      Cnt   Score    Error   Units
XmlDeserializationBenchmark.streamingXmlDeserialization                                                               thrpt        4  11.584 ±  0.132  ops/ms
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                   thrpt        4  10.740 ±  9.396  ops/ms
XmlDeserializationBenchmark.xmlDeserialization                                                                        thrpt        4  68.143 ±  2.720  ops/ms
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                            thrpt        4  68.557 ±  1.345  ops/ms
XmlDeserializationBenchmark.streamingXmlDeserialization                                                              sample   471401   0.085 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.00                            sample            0.071            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.50                            sample            0.080            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.90                            sample            0.090            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.95                            sample            0.101            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.99                            sample            0.198            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.999                           sample            0.327            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p0.9999                          sample            1.505            ms/op
XmlDeserializationBenchmark.streamingXmlDeserialization:streamingXmlDeserialization·p1.00                            sample            3.416            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization                                                  sample   468578   0.085 ±  0.001   ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.00    sample            0.074            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.50    sample            0.080            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.90    sample            0.089            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.95    sample            0.099            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.99    sample            0.200            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.999   sample            0.325            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p0.9999  sample            1.483            ms/op
XmlDeserializationBenchmark.streamingXmlIntrospectedDeserialization:streamingXmlIntrospectedDeserialization·p1.00    sample            2.834            ms/op
XmlDeserializationBenchmark.xmlDeserialization                                                                       sample   691697   0.015 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.00                                              sample            0.012            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.50                                              sample            0.014            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.90                                              sample            0.015            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.95                                              sample            0.016            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.99                                              sample            0.028            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.999                                             sample            0.066            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p0.9999                                            sample            1.238            ms/op
XmlDeserializationBenchmark.xmlDeserialization:xmlDeserialization·p1.00                                              sample            2.540            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization                                                           sample  1096537   0.018 ±  0.001   ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.00                      sample            0.016            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.50                      sample            0.017            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.90                      sample            0.018            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.95                      sample            0.023            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.99                      sample            0.032            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.999                     sample            0.076            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p0.9999                    sample            0.155            ms/op
XmlDeserializationBenchmark.xmlIntrospectedDeserialization:xmlIntrospectedDeserialization·p1.00                      sample            2.421            ms/op

Comment (jameskleeh) [reply benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java]: > With map-like deserialization, on average it takes 0.015 ms/op; while with xml mapper it is 0.020

So that is 33% slower then yeah?

Comment (svishnyakov) [reply benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java]: That is right. Though as I wrote it is not really fair comparison at this point due to incomplete implementation of `map-like first` solution. 
I am wondering how critical those 33% are, and if it worth to block addition of xml support because of it

Comment (jameskleeh) [reply benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java]: @svishnyakov No one is blocking the XML support. We've decided the way we'd like to go about doing so for good reasons (performance, less dependencies, memory consumption) and we're open to any contributing to that goal. If no one is able or willing to then we eventually will add the support. It's just a matter of priorities.

Comment (svishnyakov) [reply benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java]: @jameskleeh I do understand that and accept. That's being said I cannot completely agree with this, maybe because I don't believe xml support is going to become some of the top priorities in near future and even if it does, support of all xml trickiness is going to be really challenging and restrictive(due to required introspection) and probably incomplete support of all the legacy xml stuff like for SOAP,JAXB etc.

Comment (jameskleeh) [reply benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java]: > support of all xml trickiness is going to be really challenging and restrictive(due to required introspection) and probably incomplete support of all the legacy xml stuff like for SOAP,JAXB etc.

I don't think requiring introspection adds any challenges or restrictions. I don't know how either your proposed solution or the current direction addresses anything to do with SOAP or JAXB.

Comment (svishnyakov) [reply benchmarks/src/jmh/java/io/micronaut/deserialization/xml/XmlDeserializationBenchmark.java]: Speaking of JAXB, it is quite easy if Jackson takes cares about xml deserialization, https://github.com/FasterXML/jackson-modules-base/tree/master/jaxb; just apply a module for xml mapper and you get jaxb support

**Comment by jameskleeh** [comment]
@svishnyakoff I think I've given up on attempting to do this via streaming. We should just implement this in a way that buffers the body and binds it normally through jackson

**Comment by svishnyakoff** [comment]
@jameskleeh You mean, you might take these changes as a foundation for future xml support? Let me know if you open for further contribution in scope of xml feature and I will gladly spend my time on it

**Comment by jameskleeh** [comment]
@svishnyakoff Yeah I think we should scrap the idea of the non blocking parser because we simply can't always know the array boundaries. So the PR in its current form would need to be changed to simply buffer the XML until its complete then send it through jackson to bind to the target

**Comment by jameskleeh** [comment]
Closing this in favor of the other PR
